File: level4/bringing-a-gun-to-a-guard-fight/Solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;

public class Answer {
    
    public static int answer(int[] dimensions, int[] your_position, int[] guard_position, int distance) {
        Map<Double, Double> hitYourself = getAngles(dimensions, your_position, your_position, distance);
        Map<Double, Double> hitGuard = getAngles(dimensions, your_position, guard_position, distance);
        int solutions = 0;
        
        for (Double key : hitGuard.keySet()) {
            if (!hitYourself.containsKey(key) || hitGuard.get(key) < hitYourself.get(key)) {
                solutions++;
            }
        }
        return solutions;
    }
    
    public static double distance(int[] p1, int[] p2) {
        return Math.sqrt(Math.pow(p2[0] - p1[0], 2) + Math.pow(p2[1] - p1[1], 2));
    }
    
    public static double angle(int[] p1, int[] p2) {
        return Math.atan2(p2[1] - p1[1], p2[0] - p1[0]);
    }
    
    public static ArrayList<int[]> cartesianProduct(ArrayList<ArrayList<Integer>> series) {
        ArrayList<int[]> points = new ArrayList<>();
        
        for (int i : series.get(0)) {
            for (int j : series.get(1)) {
                points.add(new int[]{i, j});
            }
        }
        return points;
    }
    
    public static ArrayList<Integer> generateSeries(int[] dimensions, int[] your_position, int[] guard_position, int distance, int i) {
        int minimum = your_position[i] - distance;
        int maximum = your_position[i] + distance;
        
        int[] segment = {dimensions[i] - guard_position[i], guard_position[i]};
        ArrayList<Integer> series = new ArrayList<>();
        
        int current = guard_position[i];
        int segmentIndex = 0;
        while (current <= maximum) {
            series.add(current);
            current += 2 * segment[segmentIndex];
            segmentIndex = (segmentIndex + 1) % 2;
        }
        
        current = guard_position[i];
        segmentIndex = 1;
        while (current >= minimum) {
            series.add(current);
            current -= 2 * segment[segmentIndex];
            segmentIndex = (segmentIndex + 1) % 2;
        }
        
        return series;
    }
    
    public static ArrayList<ArrayList<Integer>> generateAllSeries(int[] dimensions, int[] your_position, int[] guard_position, int distance) {
        ArrayList<ArrayList<Integer>> series = new ArrayList<>();
        
        for (int i = 0; i < 2; i++) {
            series.add(generateSeries(dimensions, your_position, guard_position, distance, i));
        }
        
        return series;
    }
    
    public static Map<Double, Double> getAngles(int[] dimensions, int[] your_position, int[] guard_position, int distance) {
        Map<Double, Double> polarMap = new HashMap<>();
        
        for(int[] point : cartesianProduct(generateAllSeries(dimensions, your_position, guard_position, distance))) {
            double r = distance(point, your_position);
            double angle = angle(point, your_position);
            
            if ((point[0] != your_position[0] || point[1] != your_position[1]) && (r <= distance)) {
                polarMap.put(angle, polarMap.containsKey(angle) ? Math.min(polarMap.get(angle), r) : r);
            }
        }
        
        return polarMap;
    }
}

File: level4/bringing-a-gun-to-a-guard-fight/solution.py
import math

def answer(dimensions, your_position, guard_position, distance):
    hit_yourself = get_angles(dimensions, your_position, your_position, distance)
    hit_guard = get_angles(dimensions, your_position, guard_position, distance)
    solutions = 0
    
    for key in hit_guard:
        if key not in hit_yourself or hit_guard[key] < hit_yourself[key]:
            solutions += 1
            
    return solutions

def distance(p1, p2):
    return math.sqrt((p2[0] - p1[0]) ** 2 + (p2[1] - p1[1]) ** 2)

def angle(p1, p2):
    return math.atan2(p2[1] - p1[1], p2[0] - p1[0])

def cartesian_product(series):
    points = []
    for i in series[0]:
        for j in series[1]:
            points.append([i, j])
    return points

def generate_series(dimensions, your_position, guard_position, distance, i):
    minimum = your_position[i] - distance
    maximum = your_position[i] + distance
    segment = [dimensions[i] - guard_position[i], guard_position[i]]
    series = []
    
    current = guard_position[i]
    segment_index = 0
    while current <= maximum:
        series.append(current)
        current += 2 * segment[segment_index]
        segment_index = (segment_index + 1) % 2
    
    current = guard_position[i]
    segment_index = 1
    while current >= minimum:
        series.append(current)
        current -= 2 * segment[segment_index]
        segment_index = (segment_index + 1) % 2
    
    return series

def generate_all_series(dimensions, your_position, guard_position, distance):
    series = []
    for i in range(2):
        series.append(generate_series(dimensions, your_position, guard_position, distance, i))
    return series

def get_angles(dimensions, your_position, guard_position, distance):
    polar_map = {}
    for point in cartesian_product(generate_all_series(dimensions, your_position, guard_position, distance)):
        r = distance(point, your_position)
        angle_value = angle(point, your_position)
        if point != your_position and r <= distance:
            polar_map[angle_value] = min(polar_map.get(angle_value, float('inf')), r)
    return polar_map
